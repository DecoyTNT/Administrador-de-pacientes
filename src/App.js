import React, { Fragment, useState, useEffect } from 'react';
import Formulario from './components/Formulario';
import Cita from './components/Cita';
import PropTypes from 'prop-types';

function App() {
    let citasIniciales = JSON.parse(localStorage.getItem('citas'));
    if (!citasIniciales) {
        citasIniciales = [];
    }

    // Arreglo de citas
    const [citas, guardarCitas] = useState(citasIniciales);

    // Función que agrega una nueva cita
    const crearCita = cita => {
        guardarCitas([...citas, cita]);
    };

    // useEffect para realizar ciertas operaciones cuando el state cambia
    useEffect(() => {
        localStorage.setItem('citas', JSON.stringify(citas));
    }, [citas]);

    // Función que elimina una cita
    const eliminarCita = id => {
        const nuevasCitas = citas.filter(cita => cita.id !== id);
        guardarCitas(nuevasCitas);
    };

    return (
        <Fragment>
            <h1>Administrador de pacientes</h1>

            <div className="container">
                <div className="row">
                    <div className="one-half column">
                        <Formulario crearCita={crearCita} />
                    </div>
                    <div className="one-half column">
                        {citas.length !== 0 ? <h2>Administra tus citas</h2> : <h2>Agrega una nueva cita</h2>}

                        {citas.map(cita => (
                            <Cita key={cita.id} cita={cita} eliminarCita={eliminarCita} />
                        ))}
                    </div>
                </div>
            </div>
        </Fragment>
    );
}

Formulario.propTypes = {
    crearCita: PropTypes.func.isRequired
};

export default App;
