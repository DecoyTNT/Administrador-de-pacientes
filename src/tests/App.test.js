import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import App from './../App';

describe('Pruebas en <App />', () => {

    test('Funciona correctamente', () => {
        render(<App />);
        expect(screen.getByText('Administrador de pacientes')).toBeInTheDocument();
        expect(screen.getByTestId('nombre-app').textContent).toBe('Administrador de pacientes');
        expect(screen.getByTestId('nombre-app').tagName).toBe('H1');

        expect(screen.getByText('Crear Cita')).toBeInTheDocument();
        expect(screen.getByText('Agrega una nueva cita')).toBeInTheDocument();
    });

    test('Debe de crear una cita', () => {
        render(<App />);
        userEvent.type(screen.getByTestId('mascota'), 'Hook');
        userEvent.type(screen.getByTestId('propietario'), 'Jorge');
        userEvent.type(screen.getByTestId('fecha'), '2020-10-14');
        userEvent.type(screen.getByTestId('hora'), '11:30');
        userEvent.type(screen.getByTestId('sintomas'), 'Vomita cada que come');

        // Click en el boton de submit
        const btnSubmit = screen.getByTestId('btn-submit');
        userEvent.click(btnSubmit);

        // Revisar que no exista la alerta
        const alerta = screen.queryByTestId('alerta');
        expect(alerta).not.toBeInTheDocument();

        // Revisar por tÃ­tulo dinamuco
        expect(screen.getByTestId('titulo-dinamico').textContent).toBe('Administra tus citas');
        expect(screen.getByTestId('titulo-dinamico').textContent).not.toBe('Agrega una nueva cita');
    });

    test('Verificar las citas en el DOM', async () => {
        render(<App />);
        const citas = await screen.findAllByTestId('cita');

        expect(citas).toMatchSnapshot();

        expect(screen.getByTestId('btn-eliminar').tagName).toBe('BUTTON');
        expect(screen.getByTestId('btn-eliminar')).toBeInTheDocument();

        // Verificar una cita
        expect(screen.getByText('Hook')).toBeInTheDocument();
    });

    test('Elimina una cita', () => {
        render(<App />);

        const btnEliminar = screen.getByTestId('btn-eliminar');
        expect(btnEliminar.tagName).toBe('BUTTON');
        expect(btnEliminar).toBeInTheDocument();

        // Simular el click
        userEvent.click(btnEliminar);

        // El boton ya no debe de estar
        expect(btnEliminar).not.toBeInTheDocument();

        // No debe de estar en el documento
        expect(screen.queryByText('Hook')).not.toBeInTheDocument();
        expect(screen.queryByTestId('cita')).not.toBeInTheDocument();
    });

})
