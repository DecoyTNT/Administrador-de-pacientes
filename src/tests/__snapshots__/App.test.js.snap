// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <App /> Verificar las citas en el DOM 1`] = `
Array [
  <div
    class="cita"
    data-testid="cita"
  >
    <p>
      Mascota: 
      <span>
        Hook
      </span>
    </p>
    <p>
      Dueño: 
      <span>
        Jorge
      </span>
    </p>
    <p>
      Fecha: 
      <span>
        2020-10-14
      </span>
    </p>
    <p>
      Hora: 
      <span>
        11:30
      </span>
    </p>
    <p>
      Síntomas: 
      <span>
        Vomita cada que come
      </span>
    </p>
    <button
      class="button eliminar u-full-width"
      data-testid="btn-eliminar"
    >
      Eliminar ×
    </button>
  </div>,
]
`;
